### PR TITLE
navit: add supports, xkdb, and many fixes

### PR DESCRIPTION
**Pull request by bignaux**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/misc/navit/default.nix [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+with stdenv.lib;
 stdenv.mkDerivation rec {
   name = "navit-${version}";
   version = "0.5.1";
 
   src = fetchFromGitHub {
     owner = "navit-gps";
     repo = "navit";
-    rev = "v${version}";
+    rev = version;
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Why did you change this? Looks like they do tag their releases with `v${version}`: https://github.com/navit-gps/navit/releases

File: pkgs/applications/misc/navit/CMakeLists.txt.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 763f75b..defa74a 100755
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -212,7 +212,7 @@ CHECK_INCLUDE_FILES(endian.h HAVE_ENDIAN_H)
+ CHECK_INCLUDE_FILES(stdint.h HAVE_STDINT_H)
+ CHECK_INCLUDE_FILES(byteswap.h HAVE_BYTESWAP_H)
+ CHECK_LIBRARY_EXISTS(gypsy gypsy_control_get_default "" GYPSY_FOUND)
+-CHECK_INCLUDE_FILES(libspeechd.h HAVE_LIBSPEECHD)
```
Comment (timokau) [inline pkgs/applications/misc/navit/CMakeLists.txt.patch]: Why is this patch necessary? It should either indicate a packaging failure on our side or an insufficiently flexible build script upstream. Either way it should be fixed.

File: pkgs/applications/misc/navit/default.nix [hunk 3]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
     sha256 = "0vg6b6rhsa2cxqj4rbhfhhfss71syhnfa6f1jg2i2d7l88dm5x7d";
   };
 
-  #hardeningDisable = [ "format" ];
-  NIX_CFLAGS_COMPILE = [ "-I${SDL.dev}/include/SDL" ];
+  patches = [ ./CMakeLists.txt.patch ];
+  hardeningDisable = [ "format" ];
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Why is this necessary again?

File: pkgs/applications/misc/navit/default.nix [hunk 4]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
...
+
+  NIX_CFLAGS_COMPILE = [ ]
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Why the empty list?

File: pkgs/applications/misc/navit/default.nix [hunk 5]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
...
-  nativeBuildInputs = [ pkgconfig cmake patchelf bzip2 ];
+  nativeBuildInputs = [ makeWrapper pkgconfig cmake patchelf bzip2 ];
 
   # we dont want blank screen by defaut
   postInstall = ''
+    # copy a better configuration for nixos desktop
+    cp "${./navit.xml}" $out/share/navit/
     # emulate DSAMPLE_MAP
-    mkdir -p $out/share/navit/maps/maps
+    mkdir -p $out/share/navit/maps/
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Please quote `$out`

File: pkgs/applications/misc/navit/default.nix [hunk 6]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
...
+      qtbase qtlocation qtdeclarative qtsvg
+  ];
 
-  nativeBuildInputs = [ pkgconfig cmake patchelf bzip2 ];
+  nativeBuildInputs = [ makeWrapper pkgconfig cmake patchelf bzip2 ];
 
   # we dont want blank screen by defaut
   postInstall = ''
+    # copy a better configuration for nixos desktop
+    cp "${./navit.xml}" $out/share/navit/
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Please quote `$out/share/navit`

File: pkgs/applications/misc/navit/default.nix [hunk 7]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
...
     # emulate DSAMPLE_MAP
-    mkdir -p $out/share/navit/maps/maps
+    mkdir -p $out/share/navit/maps/
     bzcat "${sample_map}" | $out/bin/maptool "$out/share/navit/maps/osm_bbox_11.3,47.9,11.7,48.2.bin"
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Here too

File: pkgs/applications/misc/navit/default.nix [hunk 8]
```diff
@@ -21,34 +31,54 @@ stdenv.mkDerivation rec {
...
   '';
 
   # TODO: fix upstream?
   postFixup = ''
     for lib in $(find "$out/lib/navit/" -iname "*.so" ); do
-      patchelf --set-rpath ${stdenv.lib.makeLibraryPath buildInputs} $lib
+      patchelf --set-rpath ${makeLibraryPath buildInputs} $lib
     done
+    wrapProgram $out/bin/navit \
+      --prefix PATH : ${makeBinPath ([ espeak xkbd ] ++ optional speechdSupport [ speechd ]) }
```
Comment (timokau) [inline pkgs/applications/misc/navit/default.nix]: Why is `[ espeak xkbd ]` necessary now? Should that be optional?

**Review by timokau — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/navit/navit.xml [hunk 9]
```diff
@@ -0,0 +1,7324 @@
+<?xml version="1.0" encoding="UTF-8"?><!--
```
Comment (timokau) [inline pkgs/applications/misc/navit/navit.xml]: Did you submit this by accident?

File: pkgs/applications/misc/navit/default.nix (hunk 1 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/default.nix]: oh my bad!

File: pkgs/applications/misc/navit/CMakeLists.txt.patch (hunk 2 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/CMakeLists.txt.patch]: you think we should symlink the headers here like other distros ? i wonder  it was the new place and i've fix it.

File: pkgs/applications/misc/navit/default.nix (hunk 3 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/default.nix]: i don't understand well and i restore from the original navit since it was not failing.

File: pkgs/applications/misc/navit/default.nix (hunk 8 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/default.nix]: i don't make that optional since i think a gps navitation has to have it. It's forbidden to look at screen when we drive here. i can make it optional, but previously, i didn't.

File: pkgs/applications/misc/navit/navit.xml (hunk 9 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/navit.xml]: i'd like to introduce a better config for nixos.

File: pkgs/applications/misc/navit/CMakeLists.txt.patch (hunk 2 above)
Comment (timokau) [reply pkgs/applications/misc/navit/CMakeLists.txt.patch]: Are other distros doing that?

File: pkgs/applications/misc/navit/default.nix (hunk 3 above)
Comment (timokau) [reply pkgs/applications/misc/navit/default.nix]: But it was commented out previously.

File: pkgs/applications/misc/navit/default.nix (hunk 8 above)
Comment (timokau) [reply pkgs/applications/misc/navit/default.nix]: Well if it is an optional dependency according to upstream, I think it is better practice to make it optional. If you think it is essential, you can then default the optional dependency to `true`.

File: pkgs/applications/misc/navit/navit.xml (hunk 9 above)
Comment (timokau) [reply pkgs/applications/misc/navit/navit.xml]: Better than what? Does that really need a 7k line xml file? Where does that file come from?

File: pkgs/applications/misc/navit/default.nix (hunk 5 above)
Comment (Mic92) [reply pkgs/applications/misc/navit/default.nix]: Quoting is not necessary. We do not support spaces in nix store path. Even if would fix all our code we cannot fix all upstream projects. Therefore we keep code simple and do not quote at all.

File: pkgs/applications/misc/navit/default.nix (hunk 6 above)
Comment (Mic92) [reply pkgs/applications/misc/navit/default.nix]: see above

File: pkgs/applications/misc/navit/default.nix (hunk 7 above)
Comment (Mic92) [reply pkgs/applications/misc/navit/default.nix]: see above above.

File: pkgs/applications/misc/navit/navit.xml (hunk 9 above)
Comment (Mic92) [reply pkgs/applications/misc/navit/navit.xml]: We cannot accept that large files as it would affect the channel and repository size.

File: pkgs/applications/misc/navit/CMakeLists.txt.patch (hunk 2 above)
Comment (bignaux) [reply pkgs/applications/misc/navit/CMakeLists.txt.patch]: or copy https://packages.debian.org/sid/amd64/libspeechd-dev/filelist

File: pkgs/applications/misc/navit/default.nix (hunk 5 above)
Comment (timokau) [reply pkgs/applications/misc/navit/default.nix]: I didn't know that we do not support that, do you have a source on that? I agree that it would probably lead to a multitude of problems and be impractical at the moment, but that doesn't mean that we shouldn't try to do the "right" thing as much as possible.

I also disagree that leaving out the quotes makes the code simpler. It makes it necessary to reason about weather or not this variable may have spaces. It is far simpler in my opinion to always quote variables except in the cases where splitting on whitespace is explicitly wanted. Otherwise it always comes back to bite you in the future. Also quoting unconditionally would allow us to run shellcheck on scripts, as was proposed somewhere before.

So I don't see any disadvantage to quoting but quite some advantages.

File: pkgs/applications/misc/navit/CMakeLists.txt.patch (hunk 2 above)
Comment (timokau) [reply pkgs/applications/misc/navit/CMakeLists.txt.patch]: I feel like navit should be able to handle upstream behaviour. Or maybe there is some configure option for libspeechd that tells it to make those symlinks.

However if making those links is a de-facto standard among distributions, we should do it too. And we should notify upstream of that standard so that they can include the links in future releases.

File: pkgs/applications/misc/navit/default.nix (hunk 5 above)
Comment (timokau) [reply pkgs/applications/misc/navit/default.nix]: Okay, good to know. Thanks. My other points are still valid though.

Comment (Mic92) [reply pkgs/applications/misc/navit/default.nix]: At the moment omitting quotation is common sense:

```console
$ ag 'wrapProgram $out' | wc -l
596
$ ag 'wrapProgram "$out' | wc -l
179
```

There is also sometimes pitfalls of quoting twice.

If you want to change direction of how we handle quoting in builder scripts it is better to open a new issue. Because otherwise our reviews will become inconsistent which can be annoying for contributors.

Comment (timokau) [reply pkgs/applications/misc/navit/default.nix]: I don't think "the way it is commonly done" and "common sense" are necessarily the same thing.

Do do that consistently we would need a style guide and automated tooling. Thats a way bigger problem to tackle.

File: pkgs/applications/misc/navit/CMakeLists.txt.patch (hunk 2 above)
Comment (timokau) [reply pkgs/applications/misc/navit/CMakeLists.txt.patch]: I still think this should be investigated, even though Mic92 merged already.
